File: lib/pastafari_calendar_elixir_vietnamese/monster/context.ex
defmodule PastafariCalendarElixirVietnamese.Monster.Context do
  @moduledoc """
  Ngữ cảnh cơ sở của một lần gọi. Mỗi lần gọi sở hữu một giá trị riêng và không chia sẻ trạng thái ngữ nghĩa có thể thay đổi.
  """

  defstruct calculation_day: nil,
            target_day: nil,
            phase: :bootstrap,
            sub_phase: 0,
            mode: :authoritative_bootstrap,
            status: :new,
            branch_trace: [],
            metrics: %{},
            logs: [],
            diagnostics: [],
            last_error: nil
end

File: monster/context.go
package monster

import "math/big"

type Phase uint8

const (
	PhaseEntry Phase = iota
	PhaseValidate
	PhaseBootstrapStop
)

type Status uint8

const (
	StatusNew Status = iota
	StatusRunning
	StatusStoppedAtBootstrapBoundary
	StatusFailed
)

type Context struct {
	CalculationDay *big.Int
	TargetDay      *big.Int
	Phase          Phase
	Status         Status
	BranchTrace    []Phase
	Metrics        *Metrics
	Logs           *EventLog
	LastError      error
}

func NewContext(cDay, tDay *big.Int) *Context {
	return &Context{CalculationDay: new(big.Int).Set(cDay), TargetDay: new(big.Int).Set(tDay), Phase: PhaseEntry, Status: StatusNew, Metrics: NewMetrics(), Logs: NewEventLog()}
}
